### PR TITLE
Filtering - Add new option by centroid for filtering by polygon

### PR DESCRIPTION
**Pull request by Gustry**
Filtering - Add new option by centroid for filtering by polygon

The check is done when saving the editing form, when saving the CFG as well

https://github.com/3liz/qgis-lizmap-server-plugin/issues/15

### CODE REVIEW

File: lizmap/definitions/filter_by_polygon.py [hunk 1]
```diff
@@ -69,6 +75,16 @@ def __init__(self):
             'tooltip': tr('The spatial relationship to use when filtering data')
         }
 
+        self._layer_config['use_centroid'] = {
+            'type': InputType.CheckBox,
+            'header': tr('Centroid'),
+            'default': False,
+            'tooltip': tr(
+                'If the tool use the centroid of the geometry or the full geometry. It\'s quicker to use centroid. '
+                'For PostgreSQL layer, an index on the centroid is required.'
```
Comment (mdouchin) [inline lizmap/definitions/filter_by_polygon.py:84]:
```suggestion
                'If the tool must use the centroid of the geometry or the full geometry. It\'s quicker to use the centroid. '
                'For PostgreSQL layers, an index on the centroid is required.'
```

File: lizmap/definitions/filter_by_polygon.py [hunk 2]
```diff
@@ -89,6 +105,41 @@ def __init__(self):
...
+            geom=datasource.geometryColumn(),
+            )
+        )
+        if len(result) >= 1:
+            return True, None
+
+        message = tr('The layer is stored in PostgreSQL and the option "Use Centroid" is used.')
+        message += '\n'
+        message += tr(
+            'However, we could not detect a spatial index on the centroid. We must execute the given query :'
```
Comment (mdouchin) [inline lizmap/definitions/filter_by_polygon.py:133]:
```suggestion
            'However, we could not detect a spatial index on the centroid. The given query must be executed :'
```

File: lizmap/resources/ui/ui_form_filter_by_polygon.ui [hunk 3]
```diff
@@ -61,6 +61,26 @@
      </item>
     </layout>
    </item>
+   <item>
+    <widget class="QCheckBox" name="checkbox_use_centroid">
+     <property name="text">
+      <string>Use centroid of the geometry</string>
```
Comment (mdouchin) [inline lizmap/resources/ui/ui_form_filter_by_polygon.ui:67]:
```suggestion
      <string>Use the geometry centroid</string>
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,6 +7,8 @@
 * Improve the panel about server information
 * Improve metadata in the CFG file about the status of the current Lizmap Web Client version which is used
 * Improve a little the UX about the initial scales
+* Add a new option "Use centroid" for the filtering by polygon. An index on a PostgreSQL layer is required.
```
Comment (mdouchin) [inline CHANGELOG.md:10]:
```suggestion
* Add a new option "Use the geometry centroid" for the filtering by polygon. An index on the PostgreSQL layer is required.
```

**Review by mdouchin — CHANGES_REQUESTED** [review]
some minor changes suggested

File: lizmap/definitions/filter_by_polygon.py (hunk 1 above)
Comment (Gustry) [reply lizmap/definitions/filter_by_polygon.py]: I fixed all suggestions except this one.
As we are in a form, configuring a single layer, I wanted to keep it singular. I just edited with "For **a** PostgreSQL layer,..."

Hope it's alright ?

Comment (mdouchin) [reply lizmap/definitions/filter_by_polygon.py]: ok for me, thanks

**Comment by Gustry** [comment]
One step forward about having JSON booleans in the CFG instead of string.
Related to #176

**Review by mdouchin — APPROVED** [review]

**Pull request merged**